File: src/main/java/net/ndrei/teslacorelib/capabilities/TeslaCoreCapabilities.java
package net.ndrei.teslacorelib.capabilities;

import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.inventory.Slot;
import net.minecraft.nbt.NBTBase;
import net.minecraft.util.EnumActionResult;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumHand;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import net.minecraftforge.common.capabilities.Capability;
import net.minecraftforge.common.capabilities.CapabilityInject;
import net.minecraftforge.common.capabilities.CapabilityManager;
import net.ndrei.teslacorelib.capabilities.container.IGuiContainerProvider;
import net.ndrei.teslacorelib.capabilities.hud.HudInfoLine;
import net.ndrei.teslacorelib.capabilities.hud.IHudInfoProvider;
import net.ndrei.teslacorelib.capabilities.inventory.ISidedItemHandlerConfig;
import net.ndrei.teslacorelib.capabilities.wrench.ITeslaWrenchHandler;
import net.ndrei.teslacorelib.containers.BasicTeslaContainer;
import net.ndrei.teslacorelib.gui.BasicTeslaGuiContainer;
import net.ndrei.teslacorelib.gui.IGuiContainerPiece;
import net.ndrei.teslacorelib.items.TeslaWrench;

import java.util.List;

/**
 * Created by CF on 2016-12-03.
 */
public class TeslaCoreCapabilities {
    @CapabilityInject(IHudInfoProvider.class)
    public static Capability<IHudInfoProvider> CAPABILITY_HUD_INFO = null;

    @CapabilityInject(ITeslaWrenchHandler.class)
    public static Capability<ITeslaWrenchHandler> CAPABILITY_WRENCH = null;

//    @CapabilityInject(ISidedItemHandlerConfig.class)
//    public static Capability<ISidedItemHandlerConfig> CAPABILITY_SIDED_CONFIG = null;

    @CapabilityInject(ISidedItemHandlerConfig.class)
    public static Capability<IGuiContainerProvider> CAPABILITY_GUI_CONTAINER = null;

    static class CapabilityHudInfoProvider<T extends IHudInfoProvider> implements Capability.IStorage<IHudInfoProvider> {
        @Override
        public NBTBase writeNBT (Capability<IHudInfoProvider> capability, IHudInfoProvider instance, EnumFacing side) {
            return null;
        }

        @Override
        public void readNBT (Capability<IHudInfoProvider> capability, IHudInfoProvider instance, EnumFacing side, NBTBase nbt) {
        }
    }

    static class CapabilityWrenchHandler<T extends ITeslaWrenchHandler> implements Capability.IStorage<ITeslaWrenchHandler> {
        @Override
        public NBTBase writeNBT (Capability<ITeslaWrenchHandler> capability, ITeslaWrenchHandler instance, EnumFacing side) {
            return null;
        }

        @Override
        public void readNBT (Capability<ITeslaWrenchHandler> capability, ITeslaWrenchHandler instance, EnumFacing side, NBTBase nbt) {
        }
    }

//    static class CapabilitySidedConfig<T extends ISidedItemHandlerConfig> implements Capability.IStorage<ISidedItemHandlerConfig> {
//        @Override
//        public NBTBase writeNBT (Capability<ISidedItemHandlerConfig> capability, ISidedItemHandlerConfig instance, EnumFacing side) {
//            return null;
//        }
//
//        @Override
//        public void readNBT (Capability<ISidedItemHandlerConfig> capability, ISidedItemHandlerConfig instance, EnumFacing side, NBTBase nbt) {
//        }
//    }

    static class CapabilityGuiContainer<T extends IGuiContainerProvider> implements Capability.IStorage<IGuiContainerProvider> {
        @Override
        public NBTBase writeNBT (Capability<IGuiContainerProvider> capability, IGuiContainerProvider instance, EnumFacing side) {
            return null;
        }

        @Override
        public void readNBT (Capability<IGuiContainerProvider> capability, IGuiContainerProvider instance, EnumFacing side, NBTBase nbt) {
        }
    }

    static class DefaultHudInfo implements IHudInfoProvider {
        @Override
        public List<HudInfoLine> getHUDLines() {
            return null;
        }
    }

    static class DefaultWrenchHandler implements ITeslaWrenchHandler {
        @Override
        public EnumActionResult onWrenchUse(TeslaWrench wrench, EntityPlayer player, World worldIn, BlockPos pos, EnumHand hand, EnumFacing facing, float hitX, float hitY, float hitZ) {
            return EnumActionResult.PASS;
        }
    }

    static class DefaultGuiContainer implements IGuiContainerProvider {
        @Override
        public List<IGuiContainerPiece> getGuiContainerPieces(BasicTeslaGuiContainer container) {
            return null;
        }

        @Override
        public List<Slot> getSlots(BasicTeslaContainer container) {
            return null;
        }

        @Override
        public BasicTeslaContainer getContainer(int id, EntityPlayer player) {
            return null;
        }

        @Override
        public BasicTeslaGuiContainer getGuiContainer(int id, EntityPlayer player) {
            return null;
        }
    }

    public static void register() {
        CapabilityManager.INSTANCE.register(IHudInfoProvider.class, new TeslaCoreCapabilities.CapabilityHudInfoProvider<>(), DefaultHudInfo.class);
        CapabilityManager.INSTANCE.register(ITeslaWrenchHandler.class, new TeslaCoreCapabilities.CapabilityWrenchHandler<>(), DefaultWrenchHandler.class);
//        CapabilityManager.INSTANCE.register(ISidedItemHandlerConfig.class, new TeslaCoreCapabilities.CapabilitySidedConfig<>(), SidedItemHandlerConfig.class);
        CapabilityManager.INSTANCE.register(IGuiContainerProvider.class, new TeslaCoreCapabilities.CapabilityGuiContainer<>(), DefaultGuiContainer.class);
    }
}

File: src/main/kotlin/net/ndrei/teslacorelib/capabilities/TeslaCoreCapabilities.kt
package net.ndrei.teslacorelib.capabilities

import net.minecraft.entity.player.EntityPlayer
import net.minecraft.inventory.Slot
import net.minecraft.nbt.NBTBase
import net.minecraft.util.EnumActionResult
import net.minecraft.util.EnumFacing
import net.minecraft.util.EnumHand
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import net.minecraftforge.common.capabilities.Capability
import net.minecraftforge.common.capabilities.CapabilityInject
import net.minecraftforge.common.capabilities.CapabilityManager
import net.ndrei.teslacorelib.capabilities.container.IGuiContainerProvider
import net.ndrei.teslacorelib.capabilities.wrench.ITeslaWrenchHandler
import net.ndrei.teslacorelib.containers.BasicTeslaContainer
import net.ndrei.teslacorelib.gui.BasicTeslaGuiContainer
import net.ndrei.teslacorelib.gui.IGuiContainerPiece
import net.ndrei.teslacorelib.items.TeslaWrench

/**
 * Created by CF on 2017-06-28.
 */
object TeslaCoreCapabilities {
//    @CapabilityInject(IHudInfoProvider::class)
//    lateinit var CAPABILITY_HUD_INFO: Capability<IHudInfoProvider>

    @CapabilityInject(ITeslaWrenchHandler::class)
    lateinit var CAPABILITY_WRENCH: Capability<ITeslaWrenchHandler>

    @CapabilityInject(IGuiContainerProvider::class)
    lateinit var CAPABILITY_GUI_CONTAINER: Capability<IGuiContainerProvider>

//    internal class CapabilityHudInfoProvider<T : IHudInfoProvider> : Capability.IStorage<IHudInfoProvider> {
//        override fun writeNBT(capability: Capability<IHudInfoProvider>, instance: IHudInfoProvider, side: EnumFacing): NBTBase? {
//            return null
//        }
//
//        override fun readNBT(capability: Capability<IHudInfoProvider>, instance: IHudInfoProvider, side: EnumFacing, nbt: NBTBase) {}
//    }

    internal class CapabilityWrenchHandler<T : ITeslaWrenchHandler> : Capability.IStorage<ITeslaWrenchHandler> {
        override fun writeNBT(capability: Capability<ITeslaWrenchHandler>, instance: ITeslaWrenchHandler, side: EnumFacing): NBTBase? {
            return null
        }

        override fun readNBT(capability: Capability<ITeslaWrenchHandler>, instance: ITeslaWrenchHandler, side: EnumFacing, nbt: NBTBase) {}
    }

    internal class CapabilityGuiContainer<T : IGuiContainerProvider> : Capability.IStorage<IGuiContainerProvider> {
        override fun writeNBT(capability: Capability<IGuiContainerProvider>, instance: IGuiContainerProvider, side: EnumFacing): NBTBase? {
            return null
        }

        override fun readNBT(capability: Capability<IGuiContainerProvider>, instance: IGuiContainerProvider, side: EnumFacing, nbt: NBTBase) {}
    }

//    internal class DefaultHudInfo : IHudInfoProvider {
//        override val hudLines: List<HudInfoLine>
//            get() = listOf()
//    }

    internal class DefaultWrenchHandler : ITeslaWrenchHandler {
        override fun onWrenchUse(wrench: TeslaWrench, player: EntityPlayer, worldIn: World, pos: BlockPos, hand: EnumHand, facing: EnumFacing, hitX: Float, hitY: Float, hitZ: Float): EnumActionResult {
            return EnumActionResult.PASS
        }
    }

    internal class DefaultGuiContainer : IGuiContainerProvider {
        override fun getGuiContainerPieces(container: BasicTeslaGuiContainer<*>): MutableList<IGuiContainerPiece> {
            return mutableListOf()
        }

        override fun getSlots(container: BasicTeslaContainer<*>): MutableList<Slot> {
            return mutableListOf()
        }

        override fun getContainer(id: Int, player: EntityPlayer): BasicTeslaContainer<*>? {
            return null
        }

        override fun getGuiContainer(id: Int, player: EntityPlayer): BasicTeslaGuiContainer<*>? {
            return null
        }
    }

    fun register() {
//        CapabilityManager.INSTANCE.register(IHudInfoProvider::class.java, TeslaCoreCapabilities.CapabilityHudInfoProvider<IHudInfoProvider>(), DefaultHudInfo::class.java)
        CapabilityManager.INSTANCE.register(ITeslaWrenchHandler::class.java, TeslaCoreCapabilities.CapabilityWrenchHandler<ITeslaWrenchHandler>(), DefaultWrenchHandler::class.java)
        CapabilityManager.INSTANCE.register(IGuiContainerProvider::class.java, TeslaCoreCapabilities.CapabilityGuiContainer<IGuiContainerProvider>(), DefaultGuiContainer::class.java)
    }
}
